### PR TITLE
Add object detection model, dataset and training

### PR DESCRIPTION
**Pull request by seldauyanik-maxim**
Sample object detection dataset is added for street house numbers
detection problem.
Multi-box Loss is added for object detection training use.
Some utility functions for object detection models are added.
Some parameter specification for object deteciton using a yaml file
is added.
Training script is modified for object detection data loaders.
A Jupyter notebook is also added for visualization of the trained
model.

### CODE REVIEW

File: models/ai85net_adi_tinierssd.py [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+from math import sqrt
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+import ai8x
+import utils.object_detection_utils as obj_detect_utils
+
+sys.path.append('../')
```
Comment (rotx-maxim) [inline models/ai85net_adi_tinierssd.py:22]: Please ensure this is OS independent

File: models/ai85net_adi_tinierssd.py [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (rotx-maxim) [inline models/ai85net_adi_tinierssd.py:24]: This will conflict with the logic in train.py which is more complicated (e.g., supports --cpu)d

File: models/ai85net_adi_tinierssd.py [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+                        if ratio == 1.:
+                            try:
+                                additional_scale = sqrt(obj_scales[fmap] *
+                                                        obj_scales[fmaps[k + 1]])
+                            # For the last feature map, there is no "next" feature map
+                            except IndexError:
+                                additional_scale = 1.
+                            prior_boxes.append([cx, cy, additional_scale, additional_scale])
+
+        prior_boxes = torch.FloatTensor(prior_boxes).to(device)  # (1700, 4)
```
Comment (rotx-maxim) [inline models/ai85net_adi_tinierssd.py:326]: You may have to pass in `device` or a tensor on that device to get the device from, or do the conversion on the return value of this function instead.

File: parsecmd.py [hunk 4]
```diff
@@ -235,6 +235,18 @@ def get_parser(model_names, dataset_names):
     parser.add_argument('--sparsity-perf', action='store_true', default=False,
                         help='when determining best epoch, use sparsity as primary key')
 
+    obj_detection_args = parser.add_argument_group('Object Detection Arguments')
+    obj_detection_args.add_argument('--enable-obj-detection', '--obj-detection',
+                                    dest='obj_detection', default=False,
+                                    action='store_true',
+                                    help='Enable when training an Object Detection Model')
+
+    obj_detection_args.add_argument('--obj-detection-params', dest='obj_detection_params',
```
Comment (rotx-maxim) [inline parsecmd.py:244]: The `dest` is not needed here (it will convert - to _ automatically)

File: scripts/train_svhn_tinierssd.sh [hunk 5]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+cd ..
```
Comment (rotx-maxim) [inline scripts/train_svhn_tinierssd.sh:2]: Please remove `cd ..`. The scripts are expected to run as `scripts/xxxx.sh`

File: utils/object_detection_utils.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+###################################################################################################
+""" Some utility functions for Object Detection Tasks """
+
+import numpy as np
+import torch
+
+# GitHub repo for the below helper methods:
+# https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Object-Detection:
+# MIT License
+# Copyright (c) 2019 Sagar Vinodababu
```
Comment (rotx-maxim) [inline utils/object_detection_utils.py:17]: Please move this copyright to the top header, and add ours. See train.py for an example.

**Review by rotx-maxim — CHANGES_REQUESTED** [review]

**Review by rotx-maxim — CHANGES_REQUESTED** [review]
Please see comments above

File: models/ai85net_adi_tinierssd.py (hunk 1 above)
Comment (seldauyanik-maxim) [reply models/ai85net_adi_tinierssd.py:22]: Done and commit will be amended

File: models/ai85net_adi_tinierssd.py (hunk 2 above)
Comment (seldauyanik-maxim) [reply models/ai85net_adi_tinierssd.py:24]: Done and commit will be amended

File: models/ai85net_adi_tinierssd.py (hunk 3 above)
Comment (seldauyanik-maxim) [reply models/ai85net_adi_tinierssd.py:326]: Done and commit will be amended

File: scripts/train_svhn_tinierssd.sh (hunk 5 above)
Comment (seldauyanik-maxim) [reply scripts/train_svhn_tinierssd.sh:2]: Done and commit will be amended

File: parsecmd.py (hunk 4 above)
Comment (seldauyanik-maxim) [reply parsecmd.py]: Done and commit will be amended

File: utils/object_detection_utils.py (hunk 6 above)
Comment (seldauyanik-maxim) [reply utils/object_detection_utils.py:12]: Done and commit will be amended

File: losses/multiboxloss.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        bounding boxes. SSD paper uses smooth L1 loss for this purpose and so as the below
+        implementation.
+
+        Confidence Loss measures the perfroamnce of the  object classification per box.Categorical
+        cross-entropy is used to compute this loss.
+
+        SSD paper (Liu, W. et al. (2016). SSD: Single Shot MultiBox Detector,
+        https://doi.org/10.1007/978-3-319-46448-0_2)
+
+        Multi Nox Loss code is from GitHub repo:
```
Comment (rotx-maxim) [inline losses/multiboxloss.py:34]: Please move the copyright into the file header as well just like you did for the other file

**Review by rotx-maxim — CHANGES_REQUESTED** [review]

File: models/ai85net_adi_tinierssd.py [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+from math import sqrt
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+import ai8x
+import utils.object_detection_utils as obj_detect_utils
+
+sys.path.append(os.path.join(os.getcwd(), '..'))
```
[the remark this answers is not in the document]
Comment (seldauyanik-maxim) [reply models/ai85net_adi_tinierssd.py:27]: Correct, removed this one and will amend commit

File: models/ai85net_adi_tinierssd.py [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+
+        return all_images_boxes, all_images_labels, all_images_scores  # lists of length batch_size
+
+
+def tinierssd(pretrained=False, **kwargs):
+    """
+    Constructs a Tinier SSD model
+    """
+    assert not pretrained
+    # Can specify and provide different aspect ratios if required here
```
[the remark this answers is not in the document]
Comment (seldauyanik-maxim) [reply models/ai85net_adi_tinierssd.py:458]: We can, but I do not think we should, Here is why:

For a changed aspect ratio, it will be more elegant to define another method e.g. tinierssd_for_egg(), call TinierSSD constructor inside with specific aspect ratios and add another entry to the models dictionary list to make it parametrized in training script call.

If we parametrize tinierssd() instead, then inside the training script where it just constructs model with the common kwargs, it should also specify aspect ratios  and may not be the best place.

File: scripts/train_svhn_tinierssd.sh [hunk 10]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+python train.py --deterministic --print-freq 200 --pr-curves --epochs 300 --optimizer Adam --lr 0.001 --wd 0 --model tinierssd --use-bias --momentum 0.9 --weight-decay 5e-4 --dataset SVHN_74 --device MAX78000 --obj-detection --obj-detection-params parameters/obj_detection_params_svhn.yaml --batch-size 16 --qat-policy policies/qat_policy_svhn.yaml --validation-split 0 --data /data/ "$@"
```
[the remark this answers is not in the document]
Comment (seldauyanik-maxim) [reply scripts/train_svhn_tinierssd.sh:2]: Sure, removed and will be amended

File: utils/parse_obj_detection_yaml.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+            parameters = yaml.safe_load(stream)
+        except yaml.YAMLError as exc:
+            print(exc)
+
+    print(parameters)
+
+    if parameters and 'multi_box_loss' not in parameters:
+        assert False, '`multi_box_loss` parameters must be defined'
+
+    if parameters['multi_box_loss'] and 'alpha' not in parameters['multi_box_loss']:
```
[the remark this answers is not in the document]
Comment (seldauyanik-maxim) [reply utils/parse_obj_detection_yaml.py:32]: I see, the reason was the nested structure of the yaml entries and I wanted to make sure they are all set in given file. I refactored into a more readable version.

File: losses/multiboxloss.py (hunk 7 above)
Comment (seldauyanik-maxim) [reply losses/multiboxloss.py:34]: Sure done and will be commit amended

**Comment by seldauyanik-maxim** [comment]
Dear all, I will an additional small commit with a couple of additions for training in multi-GPU systems, and will re-request your reviews:
- add .module for model attribute accesses
- updated Jupyter notebook for loading saved checkpoint
I also updated requirements (related to SVHN dataset's h5py requirement for reading .mat file)

File: losses/multiboxloss.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+    """
+        Multi-box Loss for Object Detection Models
+
+        Multi-box loss is a weighted sum of the loss components: location loss and confidence loss.
+
+        Location loss  measures the distance between the annotated object boxes and the predicted
+        bounding boxes. SSD paper uses smooth L1 loss for this purpose and so as the below
+        implementation.
+
+        Confidence Loss measures the perfroamnce of the  object classification per box.Categorical
```
Comment (rotx-maxim) [inline losses/multiboxloss.py:33]: Spelling: performance

File: datasets/svhn.py [hunk 13]
```diff
@@ -0,0 +1,489 @@
...
+                if column in ['label', 'x0', 'x1', 'y0', 'y1']:
+                    self.info_df[column] = \
+                        self.info_df[column].apply(ast.literal_eval)
+        else:
+
+            mat_file_path = os.path.join(self.root_dir, self.__class__.__name__, self.d_type,
+                                         'digitStruct.mat')
+
+            if not os.path.exists(mat_file_path):
+                print('\nDownload the archieve file from: '
```
Comment (rotx-maxim) [inline datasets/svhn.py:106]: Add "Review the terms and conditions on http://ufldl.stanford.edu/housenumbers/ and then download..."

Spelling: **archive**

File: datasets/svhn.py [hunk 14]
```diff
@@ -0,0 +1,489 @@
...
+                      'http://ufldl.stanford.edu/housenumbers/[train or test].tar.gz')
+                print('Extract the downloaded archive to path [data_dir]/SVHN.')
```
Comment (rotx-maxim) [inline datasets/svhn.py:108]: Change [data_dir] to actual data directory and elaborate what the final structure is supposed to look like.

**Review by rotx-maxim — CHANGES_REQUESTED** [review]

File: losses/multiboxloss.py (hunk 12 above)
Comment (seldauyanik-maxim) [reply losses/multiboxloss.py:33]: Fixed in next commit

**Review by MaximGorkem — APPROVED** [review]
Looks great. My only comment is that we need to optimize NMS paramaters and other training parameters like qat_start to increase the accuracy. These values are the first tried ones and need to be optimized.

**Review by rotx-maxim — APPROVED** [review]
